form-flow/src/main/kotlin/com/ritense/formflow/service/FormFlowDeploymentService.kt

```diff
@@ -0,0 +1,116 @@
+import mu.KotlinLogging
+import org.everit.json.schema.loader.SchemaLoader
+import org.json.JSONObject
+import org.json.JSONTokener
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.event.EventListener
+import org.springframework.core.io.Resource
+import org.springframework.core.io.ResourceLoader
+import org.springframework.core.io.support.ResourcePatternUtils
+import org.springframework.util.StreamUtils
+import java.io.InputStream
+import java.nio.charset.StandardCharsets
+
+class FormFlowDeploymentService(
+    private val resourceLoader: ResourceLoader,
+    private val formFlowService: FormFlowService,
+) {
+
+    private val FORM_FLOW_SCHEMA_PATH = "classpath:config/form-flow/schema/formflow.schema.json"
+    private val FORM_FLOW_DEFINITIONS_PATH = "classpath:config/form-flow/*.json"
+
+    @EventListener(ApplicationReadyEvent::class)
+    fun deployAll() {
+        logger.info("Deploy all Form Flow definitions")
+        try {
+            loadFormFlowDefinitionsResources().forEach { resource ->
+                if (resource.filename != null) {
+                    deploy(resource)
+                }
+            }
+        } catch (e: Exception) {
+            throw RuntimeException("Error deploying Form Flow's", e)
+        }
+    }
+
+    fun deploy(formFlow: Resource) {
+        deploy(formFlow.filename!!.substringBeforeLast("."), formFlow.inputStream)
+    }
+
+    fun deploy(formFlowKey: String, formFlowJson: InputStream) {
+        deploy(formFlowKey, StreamUtils.copyToString(formFlowJson, StandardCharsets.UTF_8))
+    }
+
+    fun deploy(formFlowKey: String, formFlowJson: String) {
+        validate(formFlowKey, formFlowJson)
+
+        val formFlowDefinition = Mapper.get().readValue(formFlowJson, FormFlowDefinition::class.java)
+        formFlowDefinition.steps.forEach { step -> step.id.formFlowDefinition = formFlowDefinition }
+
+        try {
+            val existingDefinition = formFlowService.findLatestDefinitionByKey(formFlowKey)
+            var definitionId = FormFlowDefinitionId.newId(formFlowKey)
+
+            if (existingDefinition != null) {
+                if (formFlowDefinition == existingDefinition) {
+                    logger.info("Form Flow already deployed - {}", definitionId.toString())
+                    return
+                } else {
+                    definitionId = FormFlowDefinitionId.nextVersion(existingDefinition.id)
+                    logger.info("Form Flow changed. Deploying next version - {}", definitionId.toString())
+                }
+            }
+
+            formFlowDefinition.id = definitionId
+            formFlowService.save(formFlowDefinition)
+            logger.info("Deployed Form Flow - {}", definitionId.toString())
+        } catch (e: Exception) {
+            throw RuntimeException("Failed to deploy Form Flow $formFlowKey", e)
+        }
+    }
+
+    private fun validate(formFlowKey: String, formFlowJson: String) {
```

Comment:
Why is the key necessary if it's already in the formFlowJson?

Reply:
I get that it's done for document definitions as well, but it seems like such an arbitrary restriction. What's the point of making sure the filename is the same as the key?